File: leetcode/538.convert-bst-to-greater-tree/go/convertbst.go
package covertbst

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func convertBST(root *TreeNode) *TreeNode {
	prev := 0

	var traversal func(root *TreeNode) *TreeNode

	traversal = func(node *TreeNode) *TreeNode {
		if node == nil {
			return node
		}

		node.Right = traversal(node.Right)
		node.Val += prev
		prev = node.Val
		node.Left = traversal(node.Left)

		return node
	}

	return traversal(root)
}

File: leetcode/538.convert-bst-to-greater-tree/ts/convertbst.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const convertBST = (root: TreeNode | null): TreeNode | null => {
    let prev = 0

    const traversal = (node: TreeNode | null): TreeNode | null => {
        if (node === null) return null

        node.right = traversal(node.right)
        node.val += prev
        prev = node.val
        node.left = traversal(node.left)

        return node
    }

    return traversal(root)
}
